// xml2js/index.js
'use strict';

function stripCdata(value) {
  return value.replace(/<!\[CDATA\[([\s\S]*?)\]\]>/gi, '$1');
}

function decodeEntities(value) {
  return value
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}

function extractTag(content, tag) {
  const regex = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\/${tag}>`, 'i');
  const match = content.match(regex);
  if (!match) return undefined;
  const inner = stripCdata(match[1]).trim();
  return decodeEntities(inner);
}

function extractLink(content) {
  const attrRegex = /<link[^>]*href=['"]([^'"]+)['"][^>]*>/i;
  const attrMatch = content.match(attrRegex);
  if (attrMatch) {
    return { href: decodeEntities(attrMatch[1]) };
  }
  const value = extractTag(content, 'link');
  return value ? decodeEntities(value) : undefined;
}

function buildItem(content) {
  return {
    title: extractTag(content, 'title'),
    link: extractLink(content),
    guid: extractTag(content, 'guid'),
    pubDate: extractTag(content, 'pubDate'),
    published: extractTag(content, 'published'),
    updated: extractTag(content, 'updated'),
    description: extractTag(content, 'description')
  };
}

function parseRss(xml) {
  const items = [];
  const regex = /<item[\s\S]*?<\/item>/gi;
  let match;
  while ((match = regex.exec(xml)) !== null) {
    items.push(buildItem(match[0]));
  }
  return { rss: { channel: { item: items } } };
}

function parseAtom(xml) {
  const entries = [];
  const regex = /<entry[\s\S]*?<\/entry>/gi;
  let match;
  while ((match = regex.exec(xml)) !== null) {
    entries.push(buildItem(match[0]));
  }
  return { feed: { entry: entries } };
}

function parseStringPromise(xml) {
  return new Promise((resolve) => {
    const normalized = xml.replace(/<\?xml[^>]*>/gi, '');
    if (/<rss/i.test(normalized)) {
      resolve(parseRss(normalized));
      return;
    }
    if (/<feed/i.test(normalized)) {
      resolve(parseAtom(normalized));
      return;
    }
    resolve({});
  });
}

module.exports = { parseStringPromise };
